### ISSUE TITLE
B+ Tree Index for Extended Attributes

### ISSUE THREAD

**Issue by andreipaduroiu** [pravega/pravega#2745]
**Problem description**
The current Tier2 implementation for Extended Attributes slowly degrades in performance as the set size increases; it is estimated that beyond 1M Extended Attributes per Segment, the performance will drop so much that they would be unusable.

Current implementation:
- Tier2 sub-segment
- New updates are appended at the end
- Periodic snapshots, where all previous changes are summed up into one, organized sequence
- Reading involves reading the current snapshot + everything afterwards.

**Problem location**
SegmentAttributeIndex.java

**Suggestions for an improvement**
Implement the Tier2 Attribute file as a B+ tree with the following configuration (subject to change):
- Max Node/Page Size: 8KB
- Index Nodes: Branching factor: 256 (28 bytes per entry: 16 for key, 8 for pointer, 4 TBD)
- Leaf Nodes: Max capacity: 341 (24 bytes per entry: 16 for key, 8 for pointer, 4 TBD)

Issue #2834 has additional improvements for another time.

**Comment by tkaitchuck** [comment]
I think in practical terms 32kb may be a better number than 8kb. It for most usecases would shrink the tree depth by 1. But the cost of a network round trip returning 32kb vs 8kb is negligible, so it's probably worth the extra data transfer to save the latency of another call.

**Comment by andreipaduroiu** [comment]
Intuitively, you are correct. The bigger the page size the better. For this particular issue/PR, I chose it pretty arbitrarily.

However I have a follow-up issue #2834 where I address one of the major problems this implementation has: very fast growing index size. With that I've also written some load tests, where I compare various page sizes, batch sizes and a typical insert and random update scenario. What I've found is that, with that improvement applied (not included in this issue/PR), the optimal page size is actually 16KB. It strikes a good balance between index size and Tier2 fragmentation/IOPs.
